main.go

```diff
@@ -96,77 +114,123 @@ func logLevel() logrus.Level {
+		}
+		config.JWTTTL = jwtTTL
+	} else {
+		// The environment doesn't specify a value, use the default.
+		config.JWTTTL = jwt.TTL
 	}
+
 	// Fetch configuration data for sending emails.
-	emailURI := os.Getenv(envEmailURI)
+	config.EmailURI = os.Getenv(envEmailURI)
 	{
-		if emailURI == "" {
-			log.Fatal(envEmailURI + " is empty")
+		if config.EmailURI == "" {
+			return ServiceConfig{}, errors.New(envEmailURI + " is empty")
 		}
 		// Validate the given URI.
-		uri, err := url.Parse(emailURI)
-		if err != nil {
-			log.Fatal(errors.AddContext(err, "invalid email URI given in "+envEmailURI))
+		uri, err := url.Parse(config.EmailURI)
+		if err != nil || uri.Host == "" || uri.User == nil {
+			return ServiceConfig{}, errors.New("invalid email URI given in " + envEmailURI)
 		}
 		// Set the FROM address to outgoing emails. This can be overridden by
 		// the ACCOUNTS_EMAIL_FROM optional environment variable.
 		if uri.User != nil {
-			email.From = uri.User.Username()
+			config.EmailFrom = uri.User.Username()
 		}
 		if emailFrom := os.Getenv(envEmailFrom); emailFrom != "" {
-			email.From = emailFrom
+			config.EmailFrom = emailFrom
+		}
+		// No custom value set, use the default.
+		if config.EmailFrom == "" {
+			config.EmailFrom = email.From
 		}
 	}
 	// Fetch the configuration for maximum number of API keys allowed per user.
-	maxAPIKeys, err := strconv.Atoi(os.Getenv(envMaxNumAPIKeysPerUser))
-	if err != nil {
-		log.Printf("Warning: Failed to parse %s env var. Error: %s", envMaxNumAPIKeysPerUser, err.Error())
+	maxAPIKeysStr := os.Getenv(envMaxNumAPIKeysPerUser)
+	if maxAPIKeysStr != "" {
+		maxAPIKeys, err := strconv.Atoi(maxAPIKeysStr)
+		if err != nil {
+			log.Printf("Warning: Failed to parse %s env var. Error: %s", envMaxNumAPIKeysPerUser, err.Error())
+		}
+		if maxAPIKeys > 0 {
+			config.MaxAPIKeys = maxAPIKeys
+		}
+	} else {
+		// The environment doesn't specify a value, use the default.
+		config.MaxAPIKeys = database.MaxNumAPIKeysPerUser
 	}
-	if maxAPIKeys > 0 {
-		database.MaxNumAPIKeysPerUser = maxAPIKeys
+
+	return config, nil
+}
+
+func main() {
+	// Initialise the global context and logger. These will be used throughout
+	// the service. Once the context is closed, all background threads will
+	// wind themselves down.
+	ctx := context.Background()
+	logger := logrus.New()
+	logger.SetLevel(logLevel())
+
+	// Load the environment variables from the .env file.
+	_ = godotenv.Load()
+	config, err := parseConfiguration(logger)
+	if err != nil {
+		log.Fatal(err)
 	}
+	database.PortalName = config.PortalName
+	jwt.PortalName = config.PortalName
+	email.PortalAddressAccounts = config.PortalAddressAccounts
+	email.ServerLockID = config.ServerLockID
+	stripe.Key = config.StripeKey
+	api.StripeTestMode = config.StripeTestMode
+	jwt.AccountsJWKSFile = config.JWKSFile
+	jwt.TTL = config.JWTTTL
+	email.From = config.EmailFrom
+	database.MaxNumAPIKeysPerUser = config.MaxAPIKeys
```

Comment:
I can tell that this is going to blow up this MR so we can refactor it in a f/u but this is a similar issue as before within the parsing function. Global vars are not supposed to be used for configuration. This way you can never create two instances of an object in a thread safe way. e.g. two databases with different portal names in parallel.

You want to set the `MaxNumAPIKeysPerUser`? That should be a `staticMaxNumAPIKeysPerUser` field on the `database` instance and be passed in via something like `database.New(ctx, creds, logger, portalName, maxAPIKeysPerUser)` instead.

Reply:
https://github.com/SkynetLabs/skynet-accounts/issues/138